1-Draft/RFCNNNN-AdditionalTelemetry.md

```diff
@@ -0,0 +1,175 @@
+---
+RFC: RFC
+Author: Sydney Smith
+Status: Draft
+SupercededBy: N/A
+Version: 1.0
+Area: Telemetry
+Comments Due: 3/29/19
+Plan to implement: <Yes | No>
+---
+
+# Additional Telemetry in PowerShell
+
+Add usage telemetry across PSEdition, Platform, Application, session type and command types.
+Improve the opt-out mechanisms for PowerShell telemetry.Provide community access to aggregated
+statistics.
+
+## Motivation
+
+To make the best possible investments in PowerShell we need answers backed by data to the following questions:
+
+- Is PowerShell Core usage growing (in terms of number of starts)?
+- Is the PowerShell Core user-base growing (both in terms of number of users and devices)?
+- How is PowerShell being used? What is the usage distribution across command types and session type?
+- What are impediments to PowerShell Core usage growth?
+- What are issues that customers are hitting in PowerShell Core?
+- What versions of PowerShell tools and services should Microsoft continue to support?
+- What PowerShell integration scenarios are people using? How many people are using PowerShell
+  Editor services
+- How can we optimize the engine size and efficiency of PowerShell for cloud scenarios?
+
+To ensure we are getting an accurate picture of how everyone uses PowerShell, not just those most
+vocal/involved in the community, we need to make improvements in our telemetry.
+PowerShell usage telemetry will allow us to better prioritize testing, support, and investments.
+It should be straightforward and easy to opt-out of PowerShell telemetry (presently the process is
+not intuitive nor well documented).
+To better share our plans and successes with the community we want more complete telemetry on our
+[Public PowerBi report](https://msit.powerbi.com/view?r=eyJrIjoiYTYyN2U3ODgtMjBlMi00MGM1LWI0ZjctMmQ3MzE2ZDNkMzIyIiwidCI6IjcyZjk4OGJmLTg2ZjEtNDFhZi05MWFiLTJkN2NkMDExZGI0NyIsImMiOjV9&pageName=ReportSection5).
+Simply put, we want to make PowerShell better and believe this can be achieved by better
+understanding how PowerShell is being used.
+
+### Non-Goals
+
+The goal of this is **not** to collect any personal data of PowerShell users or infringe on
+users' privacy.
+Respecting the privacy of our users has been the principal factor in determining which metrics we
+plan to collect.
+The goal of this is not to collect patterns of individual users but to look at aggregate usage trends.
+
+## Specification
+
+Add telemetry to track the following metrics:
+
+- Count of unique devices (PowerShell)
+- Count of unique users (PowerShell)
+- Count of execution types (ex. cmdlets, native binaries/applications) (PowerShell)
+- Count of session types (ex. hosted vs not hosted sessions) (PowerShell)
+- Hosting application of the of the PowerShell session (ex. VSCode, ISE, AZDevOps) (PowerShell)
+- Version of PowerShell, PowerShellGet, PackageManagement, Nuget and platform
+  used to install from PSGallery (PowerShellGet)

```

Comment:
Additional collection suggestions:

* Name and version of any modules loaded over the course of the session, _if and only if_ they were installed via PowerShellGet -- I would also try to narrow down to modules available on the primary public PSGallery or core modules that ship in the box.
* The names of the cmdlets or functions used, perhaps subject to the same module restrictions as above?

Reply:
Gather the module/ function/ cmdlet names is very concerning to me. From a privacy and security perspective, the telemetry now can potentially gather sensitive and private information from PowerShell Core users. From the data quality perspective, when restricting to only collecting names from public PSGallery or core module, the data offers minimal value as there is already download statistics for each module in PSGallery.

Reply:
All good points, indeed. I'd forgotten about the download stats on the gallery.

Reply:
Download stats is not use stats.

Reply:
That's a good point, indeed. 🤔

Reply:
Great points. This (collecting data on specific modules from the PS Gallery) was actually something we considered but ultimately decided not to largely due to the same concerns Lawrence had. I will add a section on this in the other considerations section.